File: backend/app/schemas/metrics.py
from datetime import date, datetime

from pydantic import BaseModel


class CurrentMetrics(BaseModel):
    total_universities: int
    total_researchers: int
    total_students: int
    universities_with_tt_hardware: int
    researchers_on_tt_hardware: int
    students_on_tt_hardware: int
    last_updated: datetime | None = None


class TimelineDataPoint(BaseModel):
    date: date
    universities: int
    researchers: int
    students: int


class MetricsTimeline(BaseModel):
    data: list[TimelineDataPoint]
    start_date: date
    end_date: date


class GrowthMetrics(BaseModel):
    universities_growth: float
    researchers_growth: float
    students_growth: float
    period_days: int
    current_universities: int
    current_researchers: int
    current_students: int
    previous_universities: int
    previous_researchers: int
    previous_students: int

File: frontend/src/types/metrics.ts
export interface CurrentMetrics {
  total_universities: number;
  total_researchers: number;
  total_students: number;
  universities_with_tt_hardware: number;
  researchers_on_tt_hardware: number;
  students_on_tt_hardware: number;
  last_updated: string | null;
}

export interface TimelineDataPoint {
  date: string;
  universities: number;
  researchers: number;
  students: number;
}

export interface MetricsTimeline {
  data: TimelineDataPoint[];
  start_date: string;
  end_date: string;
}

export interface GrowthMetrics {
  universities_growth: number;
  researchers_growth: number;
  students_growth: number;
  period_days: number;
  current_universities: number;
  current_researchers: number;
  current_students: number;
  previous_universities: number;
  previous_researchers: number;
  previous_students: number;
}
